### PR TITLE
Join to template, adding python 3.12 support

### PR DESCRIPTION
**Pull request by pacrob**
Bringing the joy and wonder of the python project template to web3! (finally)

Main change is adding python 3.12 testing. 
Otherwise, most updates are structural to bring things in line with the template.

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

image

### CODE REVIEW

File: .readthedocs.yaml [hunk 1]
```diff
@@ -17,4 +17,8 @@ python:
         - docs
 
 # Build all formats for RTD Downloads - htmlzip, pdf, epub
-formats: all
+# formats: all
+# Turning off pdf for now - won't build due to long line lenghts
```
Comment (kclowes) [inline .readthedocs.yaml]: nit: 
```suggestion
# Turning off pdf for now - won't build due to long line lengths
```

File: README.md [hunk 2]
```diff
@@ -1,8 +1,10 @@
 # web3.py
 
-[![Documentation Status](https://readthedocs.org/projects/web3py/badge/?version=latest)](https://web3py.readthedocs.io/en/latest/?badge=latest)
-[![Discord](https://img.shields.io/discord/809793915578089484?color=blue&label=chat&logo=discord&logoColor=white)](https://discord.gg/GHryRvPB84)
+[![Join the conversation on Discord](https://img.shields.io/discord/809793915578089484?color=blue&label=chat&logo=discord&logoColor=white)](https://discord.gg/GHryRvPB84)
 [![Build Status](https://circleci.com/gh/ethereum/web3.py.svg?style=shield)](https://circleci.com/gh/ethereum/web3.py)
+[![PyPI version](https://badge.fury.io/py/web3.svg)](https://badge.fury.io/py/web3)
+[![Python versions](https://img.shields.io/pypi/pyversions/web3.svg)](https://pypi.python.org/pypi/web3)
+[![Docs build](https://readthedocs.org/projects/web3.py/badge/?version=latest)](https://web3.py.readthedocs.io/en/latest/?badge=latest)
```
Comment (kclowes) [inline README.md]:
```suggestion
[![Docs build](https://readthedocs.org/projects/web3.py/badge/?version=latest)](https://web3py.readthedocs.io/en/latest/?badge=latest)
```

File: .project-template/template_vars.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+web3
+web3
+web3.py
+web3.py
```
Comment (kclowes) [inline .project-template/template_vars.txt]: I believe this should be web3py (assuming it's the Doc name)

**Review by reedsa — APPROVED** [review]
Looks good, ran tests, build and docs successfully. Thanks! 🚢

File: .gitignore [hunk 4]
```diff
@@ -55,56 +57,64 @@ docs/web3.rst
...
-*.DS_Store
-
-# mypy Cache
-.mypy_cache/
-
-# pip wheel metadata
-pip-wheel-metadata/
-
-# jupyter notebook files
-*.ipynb
```
Comment (kclowes) [inline .gitignore:110]: Let's keep this in here, please :)

File: Makefile [hunk 5]
```diff
@@ -1,25 +1,29 @@
...
+	@echo "notes - consume towncrier newsfragments/ and update release notes in docs/"
+	@echo "release - package and upload a release (does not run notes target)"
 	@echo "dist - package"
 
 clean: clean-build clean-pyc
 
 clean-build:
 	rm -fr build/
 	rm -fr dist/
-	rm -fr *.egg-info
```
Comment (kclowes) [inline Makefile]: I think this should stay?

File: Makefile [hunk 6]
```diff
@@ -28,31 +32,45 @@ lint:
 	)
 
 test:
-	pytest tests
-
-test-all:
-	tox
+	python -m pytest tests
```
Comment (kclowes) [inline Makefile]: I think we want to exclude the beacon tests here? They need some special setup so I'm getting a bunch of fails. I can see the argument either way here though

File: Makefile [hunk 7]
```diff
@@ -28,31 +32,45 @@ lint:
...
 benchmark:
-	tox -e benchmark
+	python -m tox run -e benchmark
+
+autobuild-docs:
+	sphinx-autobuild --open-browser docs docs/_build/html
 
 build-docs:
-	sphinx-apidoc -o docs/ . setup.py "*conftest*" "tests" "web3/tools/*"
+	sphinx-apidoc -o docs/ . setup.py "*conftest*"
```
Comment (kclowes) [inline Makefile]: I think we want to leave this line as-is too. I'm getting a couple warnings that look like: `WARNING: autodoc: failed to import module 'tests'; the following exception was raised:
No module named 'tests'`

File: tox.ini [hunk 8]
```diff
@@ -62,24 +65,33 @@ commands=
...
 
-[testenv:py311-wheel-cli-windows]
+[testenv:windows-wheel]
 deps=
     wheel
     build[virtualenv]
 allowlist_externals=
     bash.exe
 commands=
-    bash.exe {toxinidir}/web3/scripts/release/test_windows_wheel_install.sh
```
Comment (kclowes) [inline tox.ini]: I think this should stay. See:  #3046

File: tox.ini [hunk 9]
```diff
@@ -62,24 +65,33 @@ commands=
...
 deps=
     wheel
-    build
+    build[virtualenv]
 allowlist_externals=
     /bin/rm
     /bin/bash
-    /bin/mktemp
 commands=
-    /bin/bash {toxinidir}/web3/scripts/release/test_wheel_install.sh
```
Comment (kclowes) [inline tox.ini:74]: I think this should stay. See:  #3046

**Review by kclowes — COMMENTED** [review]
This will be so nice to get in! I left a few comments, but nothing major.

File: .gitignore (hunk 4 above)
Comment (pacrob) [reply .gitignore:110]: added to template too

File: Makefile (hunk 5 above)
Comment (pacrob) [reply Makefile]: Yup. Adding to template too.

File: .gitignore (hunk 4 above)
Comment (pacrob) [reply .gitignore:110]: It's back, on line 96 now

File: setup.py [hunk 10]
```diff
@@ -5,47 +5,53 @@
...
         "pytest-mock>=1.10",
-        "pytest-watch>=4.2",
-        "pytest-xdist>=1.29",
-        "pytest>=7.0.0",
         "setuptools>=38.6.0",
-        "tox>=3.18.0",
+        "tox>=4.0.0",
         "tqdm>4.32",
-        "twine>=1.13",
+        "twine",
```
Comment (kclowes) [inline setup.py]: It probably won't be an actual problem, but we should probably keep this dep as `twine>=1.13`

**Review by kclowes — APPROVED** [review]
Looks good to me! I just left one small comment.

**Pull request merged**